File: leetcode/easy/653-two-sum-bst/iterative.cpp
/**
 * @file iterative.cpp
 * @author Vortexx2
 * @brief Problem 653 - Two Sum IV
 * @date 23-08-2021
 * 
 * Runtime - 24 ms O(n)
 * Memory Usage - 36.8 MB O(h)
 */
#include <iostream>
#include <stack>
using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
 public:
  bool findTarget(TreeNode *root, int k) {
    stack<TreeNode *> lStack, rStack;
    pushLeft(root, lStack);
    pushRight(root, rStack);

    int left = nextLeft(lStack), right = nextRight(rStack);
    while (left < right) {
      if (left + right == k) return true;
      if (left + right < k) left = nextLeft(lStack);
      else right = nextRight(rStack);
    }
    return false;
  }

  void pushLeft(TreeNode *root, stack<TreeNode *> &st) {
    while (root) {
      st.push(root);
      root = root->left;
    }
  }

  void pushRight(TreeNode *root, stack<TreeNode *> &st) {
    while (root) {
      st.push(root);
      root = root->right;
    }
  }

  int nextLeft(stack<TreeNode *> &st) {
    TreeNode *node = st.top();
    st.pop();
    pushLeft(node->right, st);
    return node->val;
  }

  int nextRight(stack<TreeNode *> &st) {
    TreeNode *node = st.top();
    st.pop();
    pushRight(node->left, st);
    return node->val;
  }
};

File: leetcode/easy/653-two-sum-bst/iterative.py
"""
  @author Vortexx2
  Problem 653 - Two Sum IV

  Runtime - 80 ms
  Memory Usage - 16.4 MB
"""
from typing import List


# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def findTarget(self, root: Optional[TreeNode], k: int) -> bool:

    def pushLeft(st: List[TreeNode], root: TreeNode):
      while root:
        st.append(root)
        root = root.left

    def pushRight(st: List[TreeNode], root: TreeNode):
      while root:
        st.append(root)
        root = root.right

    def nextLeft(st: List[TreeNode]) -> int:
      node = st.pop()
      pushLeft(st, root.right)
      return node.val

    def nextRight(st: List[TreeNode])-> int:
      node = st.pop()
      pushRight(st, root.left)
      return node.val

    lSt, rSt = [], []
    pushLeft(lSt, root)
    pushRight(rSt, root)
    left, right = nextLeft(lSt), nextRight(rSt)

    while left < right:
      if left + right == k: return True
      if left + right < k:
        left = nextLeft(lSt)
      else:
        right = nextRight(rSt)

    return False
